seldon-core-operator/tests/test_rock.py

```diff
@@ -0,0 +1,53 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+#
+# Checks out charm repository into `charm_repo`. Updated metadata.yaml with reference to locally
+# built ROCK. Executes integration tests.
+#
+
+from pathlib import Path
+
+import os
+import logging
+import pytest
+import subprocess
+import yaml
+from pytest_operator.plugin import OpsTest
+
+def read_rock_info():
+    ROCKCRAFT = yaml.safe_load(Path("rockcraft.yaml").read_text())
+    name = ROCKCRAFT["name"]
+    version = ROCKCRAFT["version"]
+    arch = list(ROCKCRAFT["platforms"].keys())[0]
+    return f"{name}_{version}_{arch}:{version}"
+
+@pytest.mark.abort_on_fail
+def test_rock(ops_test: OpsTest):
+    """Test rock."""
+    LOCAL_ROCK_IMAGE = read_rock_info()
+
+    # cleanup previous artifacts, if any
+    subprocess.run(["docker", "rm", "rockfs"])
+    subprocess.run(["rm", "-rf", "./rockfs"])
+    subprocess.run(["rm", "-rf", "./rockfs.tar"])
+
+    # create ROCK filesystem
+    subprocess.run(["docker", "create", "--name=rockfs", f"{LOCAL_ROCK_IMAGE}"])
+    subprocess.run(["docker", "export", "rockfs", "--output", "./rockfs.tar"])
+    subprocess.run(["mkdir", "./rockfs"])
+    subprocess.run(["tar", "xvf", "./rockfs.tar", "-C", "rockfs"])
+
+    # verify that all artifacts are in correct locations
+    files = os.listdir('./rockfs/opt/seldon-core-operator/')
+    assert "manager" in files
+
+    files = os.listdir('./rockfs//tmp/operator-resources/')
+    assert "service.yaml" in files
+    assert "configmap.yaml" in files
+    assert "validate.yaml" in files
+    assert "crd-v1.yaml" in files
+
+    # cleanup
+    subprocess.run(["docker", "rm", "rockfs"])
+    subprocess.run(["rm", "-rf", "./rockfs"])
+    subprocess.run(["rm", "-rf", "./rockfs.tar"])
```

Comment:
I think what would help here is to use a fixture for cleanup.  This part of the code only gets hit when the test succeeds, so any failed or errored tests will leave stuff behind.  

I haven't run this, but apart from some syntax errors I think it works:

```python
import random


@pytest.fixture()
def rock_test_env(tmpdir):
    """Yields a temporary directory and random docker container name, then cleans them up after."""
    container_name = random.choices(string.ascii, k=8)
    yield tmpdir, container_name

    try:
        subprocess.run(["docker", "rm", container_name])
    except Exception:
        pass
    # tmpdir fixture we use here should clean up the other files for us

@pytest.mark.abort_on_fail
def test_rock(ops_test: OpsTest, rock_test_env):
    """Test rock."""
    temp_dir, container_name = rock_test_env
    LOCAL_ROCK_IMAGE = read_rock_info()

    rockfs_tar = temp_dir.join("rockfs.tar")
    rockfs_dir = temp_dir.mkdir("rockfs")

    # create ROCK filesystem
    subprocess.run(["docker", "create", f"--name={container_name}", f"{LOCAL_ROCK_IMAGE}"])
    subprocess.run(["docker", "export", container_name, "--output", rockfs_tar])
    subprocess.run(["tar", "xvf", rockfs_tar, "-C", rockfs_dir])

    # verify that all artifacts are in correct locations
    files = os.listdir(rockfs_dir.join('/opt/seldon-core-operator/'))
    assert "manager" in files

    files = os.listdir(rockfs_dir.join('/tmp/operator-resources/'))
    assert "service.yaml" in files
    assert "configmap.yaml" in files
    assert "validate.yaml" in files
    assert "crd-v1.yaml" in files
```

Fixtures always complete after the test using them is done, so we'll always try to delete the docker container and leftover files.

Reply:
Awesome! Good catch!